### PR TITLE
osd: implement per-pg leases to avoid stale reads

### PR DESCRIPTION
**Pull request by liewegas**

### CODE REVIEW

File: src/messages/MOSDPing.h [hunk 1]
```diff
@@ -61,12 +61,29 @@ class MOSDPing : public Message {
   uuid_d fsid;
   epoch_t map_epoch = 0;
   __u8 op = 0;
-  utime_t stamp;
+  utime_t ping_stamp;               ///< when the PING was sent
+  ceph::signedspan mono_ping_stamp; ///< relative to sender's clock
+  ceph::signedspan mono_send_stamp; ///< replier's send stamp
+  boost::optional<ceph::time_detail::signedspan> delta_ub;  ///< ping sender
```
Comment (athanatos) [inline src/messages/MOSDPing.h]: std::optional?

**Comment by tchaikov** [comment]
retest this please.

**Comment by athanatos** [comment]
LGTM

**Comment by tchaikov** [comment]
```
/build/ceph-15.0.0-5643-ga29a4f4/src/crimson/osd/osd.cc:457:50: error: invalid new-expression of abstract class type 'ceph::osd::PG'
  457 |       return seastar::make_ready_future<Ref<PG>>(new PG{pgid,
      |                                                  ^~~~~~~~~~~~
  458 |      pg_shard_t{whoami, pgid.shard},
      |      ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~              
  459 |      coll_ref,
      |      ~~~~~~~~~                                    
  460 |      std::move(pool),
      |      ~~~~~~~~~~~~~~~~                             
  461 |      std::move(name),
      |      ~~~~~~~~~~~~~~~~                             
  462 |      create_map,
      |      ~~~~~~~~~~~                                  
  463 |      shard_services,
      |      ~~~~~~~~~~~~~~~                              
  464 |      ec_profile});
      |      ~~~~~~~~~~~                                  
In file included from /build/ceph-15.0.0-5643-ga29a4f4/src/crimson/osd/pg_map.h:14,
                 from /build/ceph-15.0.0-5643-ga29a4f4/src/crimson/osd/osd.h:27,
                 from /build/ceph-15.0.0-5643-ga29a4f4/src/crimson/osd/osd.cc:4:
/build/ceph-15.0.0-5643-ga29a4f4/src/crimson/osd/pg.h:48:7: note:   because the following virtual functions are pure within 'ceph::osd::PG':
   48 | class PG : public boost::intrusive_ref_counter<
      |       ^~
In file included from /build/ceph-15.0.0-5643-ga29a4f4/src/crimson/osd/shard_services.h:12,
                 from /build/ceph-15.0.0-5643-ga29a4f4/src/crimson/osd/osd.h:25,
                 from /build/ceph-15.0.0-5643-ga29a4f4/src/crimson/osd/osd.cc:4:
/build/ceph-15.0.0-5643-ga29a4f4/src/osd/PeeringState.h:281:18: note: 	'virtual void PeeringState::PeeringListener::queue_check_readable(epoch_t, ceph::time_detail::timespan)'
  281 |     virtual void queue_check_readable(epoch_t lpr, ceph::timespan delay) = 0;
      |                  ^~~~~~~~~~~~~~~~~~~~
/build/ceph-15.0.0-5643-ga29a4f4/src/osd/PeeringState.h:282:18: note: 	'virtual void PeeringState::PeeringListener::recheck_readable()'
  282 |     virtual void recheck_readable() = 0;
      |                  ^~~~~~~~~~~~~~~~
```

could you please help add dummy methods for crimson?

**Comment by liewegas** [comment]
@tchaikov added stubs to that it builds, but they implementations probably need to be fleshed out in order for the pg to successfully peer in many cases.

**Comment by tchaikov** [comment]
thanks @liewegas ! i will come up with a follow-up PR.

**Comment by tchaikov** [comment]
```
The following tests FAILED:
	176 - unittest_rgw_amqp (Failed)
```

https://tracker.ceph.com/issues/42042

**Comment by tchaikov** [comment]
- http://pulpito.ceph.com/kchai-2019-09-28_11:33:27-rados-wip-kefu-testing-2019-09-28-1615-distro-basic-smithi/
- http://pulpito.ceph.com/kchai-2019-09-29_02:16:32-rados-wip-kefu-testing-2019-09-28-1615-distro-basic-mira/

File: qa/suites/rados/singleton-nomsgr/all/osd_stale_reads.yaml [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+roles:
+- [mon.a, mgr.x, osd.0, osd.1, osd.2, client.0]
+overrides:
+  ceph:
+    log-whitelist:
+      - \(OSD_DOWN\)
+      - \(POOL_APP_NOT_ENABLED\)
+      - \(SLOW_OPS\)
+      - \(PG_AVAILABILITY\)
+      - \(PG_DEGRADED\)
```
Comment (tchaikov) [inline qa/suites/rados/singleton-nomsgr/all/osd_stale_reads.yaml]: http://pulpito.ceph.com/kchai-2019-09-29_02:16:32-rados-wip-kefu-testing-2019-09-28-1615-distro-basic-mira/4343377/

for posterity, the PG_DEGRADED failure was addressed by the change after this PR was picked up by my batch.

**Pull request merged**